File: core/runtime/python/xrpa_runtime/signals/signal_shared.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


from typing import TYPE_CHECKING, TypeVar, Union

from xrpa_runtime.utils.memory_accessor import (
    MemoryAccessor,
    MemoryArray,
    MemoryOffset,
    MemoryUtils,
)

if TYPE_CHECKING:
    from xrpa_runtime.signals.signal_ring_buffer import SignalRingBuffer

T = TypeVar("T")


class SignalTypeInference:
    """Utility class for inferring signal sample types."""

    @staticmethod
    def infer_sample_type(type_name: str) -> int:
        type_map = {
            "float": 0,
            "int": 1,
            "short": 2,
            "sbyte": 3,
            "uint": 4,
            "ushort": 5,
            "byte": 6,
        }

        if type_name not in type_map:
            raise ValueError(f"Unsupported sample type: {type_name}")

        return type_map[type_name]


class SignalChannelData:
    """Handles signal channel data operations."""

    def __init__(
        self,
        mem_accessor: MemoryAccessor,
        frame_count: int,
        num_channels: int,
        type_name: str,
    ):
        """
        Initialize a SignalChannelData instance.

        Args:
            mem_accessor: Memory accessor for the channel data
            frame_count: Number of frames in the channel
            num_channels: Number of channels
            type_name: Type of the sample data (float, int, etc.)
        """
        self._mem_accessor = mem_accessor
        self._frame_count = frame_count
        self._num_channels = num_channels
        self._type_name = type_name

    def get_num_channels(self) -> int:
        """Get the number of channels."""
        return self._num_channels

    def get_frame_count(self) -> int:
        """Get the number of frames."""
        return self._frame_count

    def get_channel_buffer_size(self) -> int:
        """Get the size of a single channel buffer in bytes."""
        return MemoryUtils.get_type_size(self._type_name) * self._frame_count

    def read_channel_data(
        self,
        channel_idx: int,
        dst: Union[MemoryArray, list],
        dst_start_offset: int,
        dst_count: int,
        dst_stride: int = 1,
    ) -> None:
        """
        Read data from a specific channel into a destination buffer.

        Args:
            channel_idx: Index of the channel to read from
            dst: Destination buffer (MemoryArray or list)
            dst_start_offset: Starting offset in the destination buffer
            dst_count: Number of frames to read
            dst_stride: Stride between samples in the destination buffer
        """
        src = self.access_channel_buffer(channel_idx)
        fill_count = min(self._frame_count, dst_count) if not src.is_null() else 0

        # Copy available data
        for i in range(fill_count):
            dst[dst_start_offset + i * dst_stride] = src[i]

        # Fill remaining with zeros
        for i in range(fill_count, dst_count):
            dst[dst_start_offset + i * dst_stride] = 0

    def write_channel_data(
        self, channel_idx: int, src: Union[MemoryArray, list], src_count: int
    ) -> None:
        """
        Write data to a specific channel from a source buffer.

        Args:
            channel_idx: Index of the channel to write to
            src: Source buffer (MemoryArray or list)
            src_count: Number of frames to write
        """
        dst = self.access_channel_buffer(channel_idx)
        if not dst.is_null():
            self._copy_sample_data(src, src_count, dst, self._frame_count)

    def consume_from_ring_buffer(self, ring_buffer: "SignalRingBuffer") -> None:
        """
        Consume data from a ring buffer.

        Args:
            ring_buffer: The ring buffer to consume from
        """
        channel_buffer_size = self.get_channel_buffer_size()
        out_data = self._mem_accessor.slice(
            0, channel_buffer_size * self._num_channels
        ).get_array(self._type_name)
        ring_buffer.read_deinterleaved_data(
            out_data, self._frame_count, self._frame_count
        )

    def clear_unused_channels(
        self, start_channel_idx: int, used_channel_count: int
    ) -> None:
        """
        Clear unused channels.

        Args:
            start_channel_idx: Starting channel index
            used_channel_count: Number of channels that are used
        """
        # Clear channels before the start index
        for i in range(start_channel_idx):
            dst = self.access_channel_buffer(i)
            if not dst.is_null():
                for j in range(self._frame_count):
                    dst[j] = 0

        # Clear channels after the used channels
        for i in range(start_channel_idx + used_channel_count, self._num_channels):
            dst = self.access_channel_buffer(i)
            if not dst.is_null():
                for j in range(self._frame_count):
                    dst[j] = 0

    def access_channel_buffer(self, channel_idx: int) -> MemoryArray:
        """
        Access a specific channel buffer.

        Args:
            channel_idx: Index of the channel to access

        Returns:
            MemoryArray: The channel buffer
        """
        if channel_idx < 0 or channel_idx >= self.get_num_channels():
            return MemoryArray(MemoryAccessor(), self._type_name)

        channel_buffer_size = self.get_channel_buffer_size()
        offset = channel_idx * channel_buffer_size
        return self._mem_accessor.slice(offset).get_array(self._type_name)

    def _copy_sample_data(
        self,
        src: Union[MemoryArray, list],
        src_count: int,
        dst: MemoryArray,
        dst_count: int,
    ) -> None:
        """
        Copy sample data from source to destination.

        Args:
            src: Source buffer (MemoryArray or list)
            src_count: Number of samples in source
            dst: Destination buffer
            dst_count: Number of samples in destination
        """
        copy_count = min(src_count, dst_count)

        # Copy available data
        for i in range(copy_count):
            dst[i] = src[i]

        # Fill remaining with zeros
        if src_count < dst_count:
            for i in range(src_count, dst_count):
                dst[i] = 0


class SignalPacket:
    """Represents a packet of signal data."""

    HEADER_SIZE = 16

    def __init__(self, mem_accessor: MemoryAccessor):
        """
        Initialize a SignalPacket instance.

        Args:
            mem_accessor: Memory accessor for the packet data
        """
        self._mem_accessor = mem_accessor

    def get_frame_count(self) -> int:
        """Get the number of frames in the packet."""
        offset = MemoryOffset(0)
        return self._mem_accessor.read_int(offset)

    def set_frame_count(self, frame_count: int) -> None:
        """Set the number of frames in the packet."""
        offset = MemoryOffset(0)
        self._mem_accessor.write_int(frame_count, offset)

    def get_sample_type(self) -> int:
        """Get the sample type identifier."""
        offset = MemoryOffset(4)
        return self._mem_accessor.read_int(offset)

    def set_sample_type(self, sample_type: int) -> None:
        """Set the sample type identifier."""
        offset = MemoryOffset(4)
        self._mem_accessor.write_int(sample_type, offset)

    def get_num_channels(self) -> int:
        """Get the number of channels."""
        offset = MemoryOffset(8)
        return self._mem_accessor.read_int(offset)

    def set_num_channels(self, num_channels: int) -> None:
        """Set the number of channels."""
        offset = MemoryOffset(8)
        self._mem_accessor.write_int(num_channels, offset)

    def get_frame_rate(self) -> int:
        """Get the frame rate (frames per second)."""
        offset = MemoryOffset(12)
        return self._mem_accessor.read_int(offset)

    def set_frame_rate(self, frames_per_second: int) -> None:
        """Set the frame rate (frames per second)."""
        offset = MemoryOffset(12)
        self._mem_accessor.write_int(frames_per_second, offset)

    def access_channel_data(self, type_name: str) -> SignalChannelData:
        """
        Access the channel data with the specified type.

        Args:
            type_name: Type of the sample data (float, int, etc.)

        Returns:
            SignalChannelData: The channel data
        """
        return SignalChannelData(
            self._mem_accessor.slice(self.HEADER_SIZE),
            self.get_frame_count(),
            self.get_num_channels(),
            type_name,
        )

    def copy_channel_data_from(self, src: "SignalPacket") -> None:
        """
        Copy channel data from another packet.

        Args:
            src: Source packet
        """
        self._mem_accessor.slice(self.HEADER_SIZE).copy_from(
            src._mem_accessor.slice(self.HEADER_SIZE)
        )

    @staticmethod
    def calc_packet_size(num_channels: int, sample_size: int, frame_count: int) -> int:
        """
        Calculate the size of a packet.

        Args:
            num_channels: Number of channels
            sample_size: Size of each sample in bytes
            frame_count: Number of frames

        Returns:
            int: The packet size in bytes
        """
        return SignalPacket.HEADER_SIZE + (num_channels * sample_size * frame_count)

File: core/runtime/cs/xrpa-runtime/Runtime/signals/SignalShared.cs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


using System;

namespace Xrpa
{
    public static class SignalTypeInference
    {
        public static int InferSampleType<T>()
        {
            if (typeof(T) == typeof(float))
            {
                return 0;
            }
            else if (typeof(T) == typeof(int))
            {
                return 1;
            }
            else if (typeof(T) == typeof(short))
            {
                return 2;
            }
            else if (typeof(T) == typeof(sbyte))
            {
                return 3;
            }
            else if (typeof(T) == typeof(uint))
            {
                return 4;
            }
            else if (typeof(T) == typeof(ushort))
            {
                return 5;
            }
            else if (typeof(T) == typeof(byte))
            {
                return 6;
            }
            else
            {
                throw new ArgumentException("Unsupported sample type: " + typeof(T).Name);
            }
        }
    }

    public class SignalChannelData<SampleType> where SampleType : unmanaged
    {
        private readonly MemoryAccessor _memAccessor;
        private readonly int _frameCount;
        private readonly int _numChannels;

        public SignalChannelData(MemoryAccessor memAccessor, int frameCount, int numChannels)
        {
            _memAccessor = memAccessor;
            _frameCount = frameCount;
            _numChannels = numChannels;
        }

        public int GetNumChannels()
        {
            return _numChannels;
        }

        public int GetFrameCount()
        {
            return _frameCount;
        }

        public int GetChannelBufferSize()
        {
            return MemoryUtils.GetTypeSize<SampleType>() * _frameCount;
        }

        public void ReadChannelData(int channelIdx, MemoryArray<SampleType> dst, int dstStartOffset, int dstCount, int dstStride = 1)
        {
            MemoryArray<SampleType> src = AccessChannelBuffer(channelIdx);
            int fillCount = !src.IsNull() ? Math.Min(_frameCount, dstCount) : 0;
            for (int i = 0; i < fillCount; i++)
            {
                dst[dstStartOffset + i * dstStride] = src[i];
            }
            for (int i = fillCount; i < dstCount; i++)
            {
                dst[dstStartOffset + i * dstStride] = default(SampleType);
            }
        }

        public void ReadChannelData(int channelIdx, SampleType[] dst, int dstStartOffset, int dstCount, int dstStride = 1)
        {
            MemoryArray<SampleType> src = AccessChannelBuffer(channelIdx);
            int fillCount = !src.IsNull() ? Math.Min(_frameCount, dstCount) : 0;
            for (int i = 0; i < fillCount; i++)
            {
                dst[dstStartOffset + i * dstStride] = src[i];
            }
            for (int i = fillCount; i < dstCount; i++)
            {
                dst[dstStartOffset + i * dstStride] = default(SampleType);
            }
        }

        public void WriteChannelData(int channelIdx, MemoryArray<SampleType> src, int srcCount)
        {
            MemoryArray<SampleType> dst = AccessChannelBuffer(channelIdx);
            if (!dst.IsNull())
            {
                CopySampleData(src, srcCount, dst, _frameCount);
            }
        }

        public void ConsumeFromRingBuffer(SignalRingBuffer<SampleType> ringBuffer)
        {
            int channelBufferSize = GetChannelBufferSize();
            var outData = _memAccessor.Slice(0, channelBufferSize * _numChannels).GetArray<SampleType>();
            ringBuffer.ReadDeinterleavedData(outData, _frameCount, _frameCount);
        }

        public void ClearUnusedChannels(int startChannelIdx, int usedChannelCount)
        {
            for (int i = 0; i < startChannelIdx; i++)
            {
                MemoryArray<SampleType> dst = AccessChannelBuffer(i);
                if (!dst.IsNull())
                {
                    for (int j = 0; j < _frameCount; j++)
                    {
                        dst[j] = default(SampleType);
                    }
                }
            }

            for (int i = startChannelIdx + usedChannelCount; i < _numChannels; i++)
            {
                MemoryArray<SampleType> dst = AccessChannelBuffer(i);
                if (!dst.IsNull())
                {
                    for (int j = 0; j < _frameCount; j++)
                    {
                        dst[j] = default(SampleType);
                    }
                }
            }
        }

        public MemoryArray<SampleType> AccessChannelBuffer(int channelIdx)
        {
            if (channelIdx < 0 || channelIdx >= GetNumChannels())
            {
                return (new MemoryAccessor()).GetArray<SampleType>();
            }

            int channelBufferSize = GetChannelBufferSize();
            int offset = channelIdx * channelBufferSize;
            return _memAccessor.Slice(offset).GetArray<SampleType>();
        }

        private void CopySampleData(MemoryArray<SampleType> src, int srcCount, MemoryArray<SampleType> dst, int dstCount)
        {
            int copyCount = Math.Min(srcCount, dstCount);
            for (int i = 0; i < copyCount; i++)
            {
                dst[i] = src[i];
            }

            if (srcCount < dstCount)
            {
                for (int i = srcCount; i < dstCount; i++)
                {
                    dst[i] = default(SampleType);
                }
            }
        }
    }

    public class SignalPacket
    {
        private static readonly int HeaderSize = 16;
        private readonly MemoryAccessor _memAccessor;

        public SignalPacket(MemoryAccessor memAccessor)
        {
            _memAccessor = memAccessor;
        }

        public int GetFrameCount()
        {
            var offset = new MemoryOffset(0);
            return _memAccessor.ReadInt(offset);
        }

        public void SetFrameCount(int frameCount)
        {
            var offset = new MemoryOffset(0);
            _memAccessor.WriteInt(frameCount, offset);
        }

        public int GetSampleType()
        {
            var offset = new MemoryOffset(4);
            return _memAccessor.ReadInt(offset);
        }

        public void SetSampleType(int sampleType)
        {
            var offset = new MemoryOffset(4);
            _memAccessor.WriteInt(sampleType, offset);
        }

        public int GetNumChannels()
        {
            var offset = new MemoryOffset(8);
            return _memAccessor.ReadInt(offset);
        }

        public void SetNumChannels(int numChannels)
        {
            var offset = new MemoryOffset(8);
            _memAccessor.WriteInt(numChannels, offset);
        }

        public int GetFrameRate()
        {
            var offset = new MemoryOffset(12);
            return _memAccessor.ReadInt(offset);
        }

        public void SetFrameRate(int framesPerSecond)
        {
            var offset = new MemoryOffset(12);
            _memAccessor.WriteInt(framesPerSecond, offset);
        }

        public SignalChannelData<SampleType> AccessChannelData<SampleType>() where SampleType : unmanaged
        {
            return new SignalChannelData<SampleType>(
                _memAccessor.Slice(HeaderSize), GetFrameCount(), GetNumChannels());
        }

        public void CopyChannelDataFrom(SignalPacket src)
        {
            _memAccessor.Slice(HeaderSize).CopyFrom(src._memAccessor.Slice(HeaderSize));
        }

        public static int CalcPacketSize(int numChannels, int sampleSize, int frameCount)
        {
            return HeaderSize + (numChannels * sampleSize * frameCount);
        }
    }
}
